### PR TITLE
enable phi series models.

### PR DESCRIPTION
**Pull request by lkk12014402**
# What does this PR do?
1. support phi models training
2. support static shape generation

### CODE REVIEW

**Comment by lkk12014402** [comment]
single card training with lora
image

**Comment by lkk12014402** [comment]
text-generation

python run_generation.py --model_name_or_path microsoft/phi-2 --use_kv_cache --max_new_tokens 100 --bf16

image

**Comment by lkk12014402** [comment]
hi, regisss @regisss, please review

**Comment by nprotasov** [comment]
Did you test it with 8 cards?

**Comment by lkk12014402** [comment]
> Did you test it with 8 cards?

yes, I have validated it on 4/8 cards with deepspeed zero2

**Comment by lkk12014402** [comment]
hi, regiss @regisss , will fix code with your comments today, Thanks for reviewing~

File: optimum/habana/transformers/models/phi/modeling_phi.py [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+            logger.warning_once(
+                "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`..."
+            )
+            use_cache = False
+
+    if use_cache:
+        use_legacy_cache = not isinstance(past_key_values, Cache)
+        if use_legacy_cache:
+            past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+        past_key_values_length = past_key_values.get_usable_length(seq_length)
```
[the remark this answers is not in the document]
Comment (lkk12014402) [reply optimum/habana/transformers/models/phi/modeling_phi.py:256]: yes, I add it.

File: optimum/habana/transformers/models/phi/modeling_phi.py [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+
+    if inputs_embeds is None:
+        inputs_embeds = self.embed_tokens(input_ids)
+
+    inputs_embeds = self.embed_dropout(inputs_embeds)
+
+    # Attention mask.
+    if self._use_flash_attention_2:
+        # 2d mask is passed through the layers
+        attention_mask = attention_mask if (attention_mask is not None and 0 in attention_mask) else None
```
[the remark this answers is not in the document]
Comment (lkk12014402) [reply optimum/habana/transformers/models/phi/modeling_phi.py]: removed

File: README.md [hunk 3]
```diff
@@ -181,6 +181,7 @@ The following model architectures, tasks and device distributions have been vali
 | CodeGen |   | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
 | MPT |   | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
 | Mistral |   | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
+| Phi | <div style="text-align:left"><li>DeepSpeed</li><li>LoRA</li></div>  | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
```
[the remark this answers is not in the document]
Comment (lkk12014402) [reply README.md]: yes, I have test the model with language modeling example. updated the doc!

**Comment by lkk12014402** [comment]
hi, regiss @regisss I have fixed code with your comments.

And, how can I resolve the conflicts files. It is not work by resolve the conflicts lines in my pr.

Thanks~

**Comment by regisss** [comment]
@lkk12014402 I solved the merge conflicts, it's easier that way. I'm also going to add a test for this architecture in CI.
Do you have a A100 GPU that you can access to compare the throughput with Gaudi2?

**Comment by lkk12014402** [comment]
> @lkk12014402 I solved the merge conflicts, it's easier that way. I'm also going to add a test for this architecture in CI. Do you have a A100 GPU that you can access to compare the throughput with Gaudi2?

- by using --use_hpu_graph
python run_generation.py --model_name_or_path microsoft/phi-2 --use_kv_cache --max_new_tokens 100 --bf16 --use_hpu_graphs

Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework that enables developers to build and train deep learning models at scale. It provides a set of tools and libraries that make it easy to work with large datasets and complex models.\n\nTo use DeepSpeed, you first need to install it using pip:\n\n```python\n!pip install deepspeed\n```\n\nOnce you have installed DeepSpeed, you can create a DeepSpeed engine by specifying the configuration file:\n\n```python\nimport deepspeed\n\nconfig =',)

Stats:
---------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 217.13898318691912 tokens/second
Number of HPU graphs                = 12
Memory allocated                    = 5.25 GB
Max memory allocated                = 6.09 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 7.306140886619687 seconds
---------------------------------------------------------------------------------------------------------------

- not use --use_hpu_graph

Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework that enables developers to build and train deep learning models at scale. It provides a set of tools and libraries that make it easy to work with large datasets and complex models.\n\nTo use DeepSpeed, you first need to install it using pip:\n\n```python\n!pip install deepspeed\n```\n\nOnce you have installed DeepSpeed, you can create a DeepSpeed engine by specifying the configuration file:\n\n```python\nimport deepspeed\n\nconfig =',)


Stats:
-------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 26.89582575990375 tokens/second
Number of HPU graphs                = 5
Memory allocated                    = 5.2 GB
Max memory allocated                = 5.25 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 17.410757007077336 seconds
-------------------------------------------------------------------------------------------------------------

- for A100-80G, with same setting (I revise the run_generation.py for gpu)
Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework that enables developers to build and deploy deep learning models at scale. It provides a set of tools and libraries that make it easy to train, evaluate, and deploy deep learning models.\n\nTo use DeepSpeed, we first need to install it using pip:\n\n```python\n!pip install deepspeed\n```\n\nOnce installed, we can create a DeepSpeed engine object and configure it with our model and optimizer:\n\n```python\nimport torch\nfrom',)


Stats:
----------------------------------------------------------------------
Throughput (including tokenization) = 27.234184942648884 tokens/second
----------------------------------------------------------------------

**Review by regisss — APPROVED** [review]

**Pull request merged**